app/features/signup/verify-email.ts

```diff
@@ -0,0 +1,65 @@
+import { useEffect, useRef, useState } from 'react';
+import { useUserStore } from '~/features/user/user-provider';
+import { useToast } from '~/hooks/use-toast';
+
+export const useVerifyEmailOnLoad = ({
+  code,
+  onFailed,
+}: {
+  code: string | undefined;
+  onFailed: () => void;
+}) => {
+  const verifyEmail = useUserStore((state) => state.verifyEmail);
+  const { toast } = useToast();
+  const failedRef = useRef(onFailed);
+
+  useEffect(() => {
+    if (code) {
+      const handleVerificationCode = async () => {
+        try {
+          await verifyEmail(code);
+        } catch {
+          toast({
+            variant: 'destructive',
+            title: 'Failed to verify',
+            description:
+              'Please try entering the code manually or request another verification email',
+          });
+          failedRef.current();
+        }
+      };
+      handleVerificationCode();
+    }
+  }, [code, verifyEmail, toast]);
+};
+
+export const useRequestEmailVerificationCode = () => {
```

Comment:
Should we make some delay so that the user has to wait 30 seconds before requesting a new code? Fine for now, just something to possibly consider for the future. Not sure if its important or not, but seems like a common practice I've observed